### PR TITLE
[Bugfix] Rescale NVFP4 weight scales to fix BF16 dequant underflow

### PR DESCRIPTION
**Pull request by ricky-chaoju**
## Summary

The Marlin NVFP4 kernel converts weight scales to a special S0E5M3 format by multiplying by `2^7`, assuming the MSB is always 1 for non-zero values. However, some models (e.g., those with `global_scale=1`) produce small scale values that violate this assumption. On BF16, the exponent widening in `dequant_fp8_scales` interprets these as ~2^(-112), causing underflow to zero and garbled output.

**Fix:** Rescale weight scales in Python by a power-of-2 factor so all non-zero values satisfy the MSB=1 assumption after `* 2^7`, then compensate by dividing `global_scale` accordingly. This avoids any kernel modification.

## Changes

- Add `_nvfp4_compute_scale_factor` to compute the rescaling factor  
- Apply rescaling in `nvfp4_marlin_process_scales` and compensate in `global_scale`  
- Support shared scale factor across experts for MoE layers  

## Test Plan

Tested with [GLM-4.7-Flash-NVFP4](https://huggingface.co/GadflyII/GLM-4.7-Flash-NVFP4) on RTX 4090:

- **BF16:** GSM8K 5-shot accuracy = 81.3%, 0% invalid (was garbled output before fix)

Related #34694

### CODE REVIEW

**Comment by mgoin** [comment]
The obvious issue with this is the impact on performance due to needing to dynamically generate scales for activations and apply them, so a few base questions:
1. What is the need for running modern models (that are trained in bfloat16) with `--dtype float16`, even with NVFP4? Like you've stated in the code, we don't need this sort of correction generally when using bfloat16 and the default dtype for most models of interest will be bfloat16
2. If the issue is just intermediate values, why not look at changing the accumulation strategy in the kernel for float16? Maybe because float32 accumulation is slow on consumer GPUs like RTX 4090?
3. Can you show the quality issue (and performance impact) with a structured eval? The easiest one is to use the gsm8k eval we have in `python tests/evals/gsm8k/gsm8k_eval.py` that works with the server i.e. `vllm serve your-model`

**Comment by ricky-chaoju** [comment]
Thanks for the review! After investigating more deeply, I found a pre-existing kernel bug that was causing completely garbled output for all
BF16 NVFP4 models on the Marlin fallback path — `dequant_fp8_scales<nv_bfloat162>` had incorrect exponent widening from S0E5M3 to BF16, causing
most scale values to underflow to zero. This is now fixed in the second commit.

To address your questions:

> What is the need for running modern models with --dtype float16?

You're right that BF16 is the recommended path — and this PR fixes BF16 which was silently broken. The FP16 stability scaling is for older GPUs
like V100 that don't support BF16 at all.

> If the issue is just intermediate values, why not change the accumulation strategy?

The FP16 overflow happens during weight dequantization (FP4 → FP16), before accumulation. FP32 accumulation is already used in the MMA. The
stability scaling addresses the pre-accumulation intermediate values.

> Can you show the quality issue with a structured eval?

GSM8K 5-shot on GLM-4.7-Flash-NVFP4 (RTX 4090):

| | Before this PR | After this PR |
|---|---|---|
| BF16 | Garbled output (scale dequant bug) | **78.0% accuracy**, 0% invalid |

FP16 single-layer kernel verified correct (cos_sim = 1.000 vs FP32 ref).
FP16 stability scaling is a safety net for models with larger scale values
where intermediate overflow may occur — K = 1.0 (no-op) when no risk detected.


The BF16 fix is the critical change — exponent bit manipulation in `dequant.h` that was silently breaking all NVFP4 models on non-native-FP4
GPUs. The FP16 stability scaling is a safety net that activates only when overflow risk exists (K > 1.0), with zero overhead otherwise.

**Comment by mgoin** [comment]
Cc @jinzhen-lin to check the bf16 math. I still need to work it out myself to feel confident it won't cause regression

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py [hunk 1]
```diff
@@ -27,7 +27,24 @@ def is_fp4_marlin_supported():
...
+    nonzero_mask = ws_float > 0
+    if nonzero_mask.any():
+        min_val = ws_float[nonzero_mask].min()
+        if min_val < 2**-6:
+            sf = ((2**-6) / min_val).log2().ceil().exp2()
+            assert (ws_float[nonzero_mask] * sf <= 448).all(), (
+                "NVFP4 scale dynamic range too large for rescaling"
+            )
+            return sf
+    return torch.tensor(1.0)
```
Comment (jinzhen-lin) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: I suggest returning a Python float (`sf.item()` and `1.0`) directly instead of a Tensor. This avoids the need to manually manage the tensor's device and dtype later on.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py [hunk 2]
```diff
@@ -27,7 +27,24 @@ def is_fp4_marlin_supported():
...
-def nvfp4_marlin_process_scales(marlin_scales):
+def _nvfp4_compute_scale_factor(marlin_scales_half):
+    """Compute the power-of-2 scale_factor needed so that all non-zero
+    values in marlin_scales_half are >= 2^-6 after rescaling.
+    Returns a scalar torch.Tensor (power of 2, >= 1.0)."""
+    ws_float = marlin_scales_half.float()
+    nonzero_mask = ws_float > 0
+    if nonzero_mask.any():
+        min_val = ws_float[nonzero_mask].min()
+        if min_val < 2**-6:
```
Comment (jinzhen-lin) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: Since you're rescaling in `nvfp4_marlin_process_scales`, why not do `scales * (2 ** 7)` before computing the `scale_factor`? It makes the `min_val < 2` check much easier to understand.

**Review by jinzhen-lin — APPROVED** [review]
Thank you. The new fix LGTM. Just some more suggestions.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py (hunk 1 above)
Comment (ricky-chaoju) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: Good point, done. `_nvfp4_compute_scale_factor` now returns `sf.item() / 1.0` directly, and I removed all `.to(device, dtype)` at the call sites.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py (hunk 2 above)
Comment (ricky-chaoju) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: Done! Moved the `* (2 ** 7)` into `_nvfp4_compute_scale_factor`, so the threshold is now `min_val < 2` instead of `min_val < 2 ** -6`.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py [hunk 3]
```diff
@@ -27,7 +27,24 @@ def is_fp4_marlin_supported():
...
+        if min_val < 2:
+            sf = (2 / min_val).log2().ceil().exp2()
+            assert (ws_float[nonzero_mask] * sf <= 448 * (2**7)).all(), (
+                "NVFP4 scale dynamic range too large for rescaling"
+            )
+            return sf.item()
+    return 1.0
+
+
+def nvfp4_marlin_process_scales(marlin_scales, scale_factor=None):
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: nit: I notice many of the functions in this file do not use type annotations on the arguments, could you add these?

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py [hunk 4]
```diff
@@ -408,6 +443,11 @@ def prepare_moe_fp4_layer_for_marlin(
         else:
             size_n, size_k = k, n
 
+        # For NVFP4: compute unified scale_factor across all experts
+        combined_scale_factor = None
+        if is_nvfp4:
+            combined_scale_factor = _nvfp4_compute_scale_factor(scales.to(torch.half))
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: What is the purpose of casting to torch.half just to have it upconverted to torch.float within the function?

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py (hunk 3 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: Can you also add a docstring to the fucntion to describe the purpose of scale_factor and how it is used if passed in vs not?

Comment (ricky-chaoju) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: Done ,added type annotations and a docstring in the latest push.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py (hunk 4 above)
Comment (ricky-chaoju) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: The .to(torch.half) was unnecessary since _nvfp4_compute_scale_factor immediately converts to .float() internally. Removed in the latest push.

**Review by mgoin — APPROVED** [review]
